niceman/interface/exec.py

```diff
@@ -32,6 +33,211 @@
 lgr = getLogger('niceman.api.exec')
 
 
+class CommandAdapter(object):
+    """Interface to kinds of commands `exec` can execute.
+
+    Each subclass should define a `execute` method and optionally defines
+    `pre_command` or `post_command` to perform pre- or post-command operations.
+    The caller should execute the command by calling the instance object
+    itself.
+
+    Parameters
+    ----------
+    resource : Resource object
+    command : str
+        The program to run.
+    cmd_args : list of str
+        Arguments to pass to `command`.
+    """
+
+    def __init__(self, resource, command, cmd_args):
+        self.resource = resource
+        self.session = resource.get_session()
+        self.command = command
+        self.cmd_args = cmd_args
+
+    def pre_command(self):
+        pass
+
+    def execute(self):
+        """Run the full command defined by `command` and `cmd_args`.
+
+        Returns
+        -------
+        A tuple, (stdout, stderr).
+        """
+        raise NotImplementedError
+
+    def post_command(self):
+        pass
+
+    def __call__(self):
+        self.pre_command()
+        try:
+            out, err = self.execute()
+        finally:
+            self.post_command()
+        return out, err
+
+
+class PlainCommand(CommandAdapter):
+
+    def execute(self):
+        return self.session.execute_command([self.command] + self.cmd_args)
+
+
+class InternalCommand(CommandAdapter):
+
+    def execute(self):
+        self.session.niceman_exec(self.command, self.cmd_args)
+        return None, None  # Return same form as Session.execute_command.
+
+
+class TracedCommand(CommandAdapter):
+    """A tracer-wrapped command.
+
+    In the pre-command step, a tracer (extracted from Reprozip) is downloaded
+    to the local machine if not already present.  This tracer is transferred to
+    the resource on which the command will be executed.
+
+    After the command is executed under the tracer, the post-command step
+    downloads the trace artifacts locally and calls `niceman retrace` on the
+    Reprozip spec.
```

Comment:
May be elaborate just a bit more, that with such setup reprozip is required only on the local host and not on the remote machine (if session is remote)?

Reply:
will do